Ensure all artifact symbols are defined before emitting object

pchickey opened the pull request:
Fixes #30 

Adds a new method `Artifact::undefined_symbols` that gives the set of non-import declarations that have no definitions. Ensures that this is an empty set in the `emit` method.

Additionally, prevents duplicate definitions of symbols.

sunfishcode commented on src/artifact.rs:
```diff
@@ -447,10 +447,34 @@ impl Artifact {
+                Decl::FunctionImport => {},
+                Decl::DataImport => {},
+                Decl::Function { .. } | Decl::Data { .. } | Decl::CString { .. }=> {
+                    if self.definitions.iter().find(|def| def.name == name).is_none() {
```
What would you think about adding a `defined` flag to the `Function`/`Data`/`CString` arms of `Decl`, and having the `define` function set it to true? That way we wouldn't have to scan the definitions for each import here.

sunfishcode commented on src/artifact.rs:
```diff
@@ -447,10 +447,34 @@ impl Artifact {
 
+    /// Get set of non-import declarations that have not been defined. This must be an empty set in
+    /// order to `emit` the artifact.
+    pub fn undefined_symbols(&self) -> Vec<String> {
```
At first I was worried that this was going to allocate a `Vec` of allocated `String` all the time, but it turns out that when everything is working as intended, the `Vec` is empty, so nothing is allocated :-).

pchickey commented:
I thought of a way to reduce the time complexity of undefined_symbols at the cost of cloning the declarations, which I think is worth it? (I would not want to clone the definitions.)

sunfishcode commented on src/artifact.rs:
```diff
@@ -447,10 +447,36 @@ impl Artifact {
+        let mut decls = self.declarations.clone();
+        for def in self.definitions.iter() {
+            decls.remove(&def.name);
+        }
```
This is algorithmically better :-), but I'm curious if there's a reason for preferring this over adding `defined` flags to `Decl`? 

A `defined` flag in `Decl` could also be used to diagnose a symbol being defined multiple times too, which seems like it would be nice.

sunfishcode commented on src/artifact.rs:
```diff
@@ -447,10 +447,36 @@ impl Artifact {
+        for (&name, &decl) in decls.iter() {
+            match decl {
+                Decl::FunctionImport | Decl::DataImport => {},
+                Decl::Function { .. } | Decl::Data { .. } | Decl::CString { .. } => {
```
I just noticed that `Decl` has an `is_import()` function, which would tidy things up here.

pchickey commented:
Rebased on #33; changed internal representation of Decl to have a flag indicating whether it is defined, which brings the search down to O(n), and allows us to throw an error in Artifact::define for duplicate definitions.
